File: python/hist.py
from shiny import App, module, reactive, render, ui
import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import gaussian_kde
    
@module.server
def hist_server(input, output, session):
    """Hist Server."""
    from rpy2 import robjects

    # `faithful$eruptions` from R
    eruptions = robjects.r['faithful'][0]
    
    @render.plot
    def main_plot():
        fig, ax = plt.subplots()
        n_breaks = int(input.n_breaks())
        hist_data = np.histogram(eruptions, bins=n_breaks, density=True)
        ax.bar(hist_data[1][:-1], hist_data[0], width=np.diff(hist_data[1]), edgecolor='black', align='edge')

        if input.individual_obs():
            ax.plot(eruptions, np.zeros_like(eruptions), 'r|', markersize=10)

        if input.density():
            bw_adjust = input.bw_adjust()
            kde = gaussian_kde(eruptions, bw_method=bw_adjust)
            x_grid = np.linspace(min(eruptions), max(eruptions), 1000)
            ax.plot(x_grid, kde(x_grid), color='blue')

        ax.set_xlabel("Duration (minutes)")
        ax.set_title("Geyser eruption duration")
        
    @render.ui
    def output_bw_adjust():
        if input.density():
            return ui.input_slider(
                "bw_adjust",
                "Bandwidth adjustment:",
                min=0.2,
                max=2.0,
                value=1.0,
                step=0.2
            )
    
# hist_server("hist")

@module.ui
def hist_input():
    """Hist Input."""
    return ui.card(
        ui.input_select(
            "n_breaks",
            "Number of bins in histogram (approximate):",
            choices=[10, 20, 35, 50],
            selected=20
        ),
        ui.input_checkbox(
            "individual_obs",
            "Show individual observations",
            value=False
        ),
        ui.input_checkbox(
            "density",
            "Show density estimate",
            value=False
        )
    )

# hist_input("hist")

@module.ui
def hist_output():
    """Hist Output."""
    return ui.output_plot("main_plot")

# hist_output("hist")


@module.ui
def hist_ui():
    """Hist UI."""
    return ui.output_ui("output_bw_adjust")

def hist_app():
    """Hist App."""
    app_ui = ui.page_fluid(
        hist_input("hist"),
        hist_output("hist"),
        hist_ui("hist")
    )
    def app_server(input, output, session):
        hist_server("hist")

    app = App(app_ui, app_server)

    #if __name__ == '__main__':
    app.run()

# hist_app()

File: R/hist.R
#' Shiny Server for Geyser Graphics Histogram
#'
#' @param id shiny identifier
#' @param df reactive data frame

#' @return reactive server
#' @export
#' @rdname histServer
#' @importFrom shiny checkboxInput moduleServer NS plotOutput renderPlot 
#'             renderUI selectInput shinyApp sliderInput uiOutput
#' @importFrom bslib page
#' @importFrom graphics hist lines rug
#' @importFrom stats density
#' @importFrom stringr str_to_title
histServer <- function(id, df = shiny::reactive(faithful)) {
  shiny::moduleServer(id, function(input, output, session) {
    ns <- session$ns
    
    # Output Main Plot
    output$main_plot <- shiny::renderPlot({
      shiny::req(df())
      if(ncol(df()) < 1) return(NULL)
      
      xvar <- colnames(df())[1]
      graphics::hist(df()[[xvar]],
                     probability = TRUE,
                     breaks = as.numeric(input$n_breaks),
                     xlab = xvar,
                     main = stringr::str_to_title(xvar))
      
      if (input$individual_obs) {
        graphics::rug(df()[[xvar]])
      }
      if (input$density) {
        shiny::req(df())
        shiny::req(input$bw_adjust)
        dens <- stats::density(df()[[xvar]],
                               adjust = input$bw_adjust)
        graphics::lines(dens, col = "blue")
      }
    })
    
    # Input Bandwidth Adjustment
    output$bw_adjust <- shiny::renderUI({
      if(input$density) {
        shiny::sliderInput(inputId = ns("bw_adjust"),
                           label = "Bandwidth adjustment:",
                           min = 0.2, max = 2, value = 1, step = 0.2)
      }
    })
  })
}
#' Shiny Module Input for Geyser Graphics Histogram
#' @param id identifier for shiny reactive
#' @return nothing returned
#' @rdname histServer
#' @export
histInput <- function(id) {
  ns <- shiny::NS(id)
  shiny::tagList(
    shiny::selectInput(inputId = ns("n_breaks"),
                label = "Number of bins in histogram (approximate):",
                choices = c(10, 20, 35, 50),
                selected = 20),
    shiny::checkboxInput(inputId = ns("individual_obs"),
                  label = shiny::strong("Show individual observations"),
                  value = FALSE),
    
    shiny::checkboxInput(inputId = ns("density"),
                  label = shiny::strong("Show density estimate"),
                  value = FALSE))
}
#' Shiny Module UI for Geyser Graphics Histogram
#' @param id identifier for shiny reactive
#' @return nothing returned
#' @rdname histServer
#' @export
histUI <- function(id) {
  ns <- shiny::NS(id)
  shiny::uiOutput(ns("bw_adjust"))
}
#' Shiny Module Output for Geyser Graphics Histogram
#' @param id identifier for shiny reactive
#' @return nothing returned
#' @rdname histServer
#' @export
histOutput <- function(id) {
  ns <- shiny::NS(id)
  shiny::plotOutput(ns("main_plot"), height = "300px")
}
#' Shiny Module App for Geyser Graphics Histogram
#' @return nothing returned
#' @rdname histServer
#' @export
histApp <- function() {
  ui <- bslib::page(
    histInput("hist"), 
    histOutput("hist"),
    # Display this only if the density is shown
    histUI("hist")
  )
  server <- function(input, output, session) {
    histServer("hist") 
  }
  shiny::shinyApp(ui, server)
}
